src/common/dout_fmt.h

```diff
@@ -0,0 +1,49 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright contributors to the Ceph project
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#pragma once
+
+#include <iosfwd>
+#include <iterator>
+#include <fmt/core.h>
+#include "dout.h"
+
+// dout macros that use libfmt instead of ostream-style formatting
+
+// work around "warning: value computed is not used" with default dout_prefix
+inline void dout_fmt_use_prefix(std::ostream&) {}
+
+#define lsubdout_fmt(cct, sub, v, ...) \
+  dout_impl(cct, ceph_subsys_##sub, v) \
+  dout_fmt_use_prefix(dout_prefix); \
+  fmt::format_to(std::ostream_iterator<char>(*_dout), __VA_ARGS__); \
+  *_dout << dendl
+
+#define ldout_fmt(cct, v, ...) \
+  dout_impl(cct, dout_subsys, v) \
+  dout_fmt_use_prefix(dout_prefix); \
+  fmt::format_to(std::ostream_iterator<char>(*_dout), __VA_ARGS__); \
+  *_dout << dendl
+
+#define dout_fmt(v, ...) \
+  ldout_fmt((dout_context), v, __VA_ARGS__)
```

Comment:
This looks really great, much better than building a debug line piecemeal. Shall we also mention in a comment that this can work for all Ceph types that implement:

https://fmt.dev/latest/api.html#std-ostream-support

Also note:

https://github.com/ceph/ceph/blob/c80f7fd3f9a888f6333ceecb58abe43752711bac/src/common/dout.h#L47-L57

With some templates, let's add:

```
template<class T> requires HasPrint<T>
struct fmt:formatter<T> : ostream_formatter {};
```

? (Haven't tried this but I think it should work.)

Reply:
really appreciate it @cbodley

Reply:
I'm not sure we should encourage relying in existing ostream support (which is slow).
The best option is to always add a formatter to each new type. And common_fmt adds a lazy option, too.

Reply:
thanks for the reviews! i think i'd prefer to limit the pr's scope just to dout, rather than the formatting of ceph types in general. we're using fmt quite a bit already, and a lot of our types already have formatters

this pr was motivated by discussion in https://github.com/ceph/ceph/pull/55076#discussion_r1587975786 and https://github.com/ceph/ceph/pull/56008#pullrequestreview-1944064024, where the combination of libfmt and std::ostream led to unnecessary overhead. we'll get the best performance if we stick to one model or the other

i tend to agree with @ronen-fr that their ostream_formatter is probably adding too much overhead to be a good default. i see from [detail::format_value()](https://github.com/fmtlib/fmt/blob/master/include/fmt/ostream.h#L98) that it's constructing a `std::basic_ostream` and `std::basic_streambuf` for each value. long ago in https://github.com/ceph/ceph/pull/19100 we found that those constructor calls were surprisingly expensive (i think because of a global lock to look up the locale?)

it's fine for types to opt in to the ostream_formatter on a case-by-case basis, but the lack of default ostream support will at least make us stop and think about it. if there are enough examples of good custom formatters, we'll be more likely to copy one of those instead

do you buy that @batrick?

Reply:
> I'm not sure we should encourage relying in existing ostream support **(which is slow)**.

huh?

As I understand things, this PR is only optimizing the case where we'd like to use `fmt::format` to create an intermediate formatted string (at least one extra heap allocation) when we could just write to the `ostream`?

Reply:
>long ago in https://github.com/ceph/ceph/pull/19100

it's changed a lot, more recently than that: #23721

I put a lot of effort into eliminating all heap allocations for debug logging. (So the "slow" comment with respect to the use of the `ostream` is somewhat puzzling.)

> do you buy that @batrick?

Maybe I misunderstood what

`struct fmt:formatter<T> : ostream_formatter {};`

was actually doing. My hope was that it would teach `fmt::format_to` to call our `T::print` method with the `ostream` we passed. If it's creating an intermediate ostream then that sucks and makes it unsuitable.

Reply:
> If it's creating an intermediate ostream then that sucks and makes it unsuitable.

that's my understanding. this pr passes a `std::ostream_iterator` to `fmt::format_to(OutputIt)`, so fmt never actually has access to our dout stream. and the same formatters have to work with `fmt::format()` where there is no underlying ostream

Reply:
Circling back: `fmt::print` is pretty sophisticated about this:

```
$ cat fmt-ostream.cc 
#include <iostream>
#include <fmt/core.h>
#include <fmt/ostream.h>
#include <sstream>


struct dumpme {
  int a;
  int b;
  friend inline std::ostream& operator<<(std::ostream& os, const dumpme& o) {
    return os << "long string to heap allocate: " <<  o.a << " " << o.b;
  }
};

template<> struct fmt::formatter<dumpme> : ostream_formatter {};

int main (int argc, char*argv[])
{
  auto o = dumpme{1,2};

  auto opt = std::string(argv[1]);

  for (int i = 0; i < 1000; ++i) {
    if (opt == "dump") {
      std::cout << "dumpme is " << o << std::endl;
    } else if (opt == "string") {
      std::string s = fmt::format("dumpme is {}\n", o);
      std::cout << s;
    } else if (opt == "print") {
      fmt::print(std::cout, "dumpme is {}\n", o);
    }
  }

  return 0;
}
```

running:

```
$ g++ --std=c++20 fmt-ostream.cc -lfmt ; for s in dump string print; do valgrind --leak-check=full ./a.out $s > /dev/null; done
==1402093== Memcheck, a memory error detector
==1402093== Copyright (C) 2002-2024, and GNU GPL'd, by Julian Seward et al.
==1402093== Using Valgrind-3.23.0 and LibVEX; rerun with -h for copyright info
==1402093== Command: ./a.out dump
==1402093== 
==1402093== 
==1402093== HEAP SUMMARY:
==1402093==     in use at exit: 0 bytes in 0 blocks
==1402093==   total heap usage: 2 allocs, 2 frees, 77,824 bytes allocated
==1402093== 
==1402093== All heap blocks were freed -- no leaks are possible
==1402093== 
==1402093== For lists of detected and suppressed errors, rerun with: -s
==1402093== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
==1402101== Memcheck, a memory error detector
==1402101== Copyright (C) 2002-2024, and GNU GPL'd, by Julian Seward et al.
==1402101== Using Valgrind-3.23.0 and LibVEX; rerun with -h for copyright info
==1402101== Command: ./a.out string
==1402101== 
==1402101== 
==1402101== HEAP SUMMARY:
==1402101==     in use at exit: 0 bytes in 0 blocks
==1402101==   total heap usage: 1,002 allocs, 1,002 frees, 122,824 bytes allocated
==1402101== 
==1402101== All heap blocks were freed -- no leaks are possible
==1402101== 
==1402101== For lists of detected and suppressed errors, rerun with: -s
==1402101== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
==1402108== Memcheck, a memory error detector
==1402108== Copyright (C) 2002-2024, and GNU GPL'd, by Julian Seward et al.
==1402108== Using Valgrind-3.23.0 and LibVEX; rerun with -h for copyright info
==1402108== Command: ./a.out print
==1402108== 
==1402108== 
==1402108== HEAP SUMMARY:
==1402108==     in use at exit: 0 bytes in 0 blocks
==1402108==   total heap usage: 2 allocs, 2 frees, 77,824 bytes allocated
==1402108== 
==1402108== All heap blocks were freed -- no leaks are possible
==1402108== 
==1402108== For lists of detected and suppressed errors, rerun with: -s
==1402108== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

There are no intermediate allocations (i.e. an intermediate `std::ostream` created by `fmt::print`). So I think teaching `fmt::print` to handle our types with `operator<<` or `requires HasPrint` would be a step in the right direction with no loss in performance.

Reply:
the code i linked above https://github.com/fmtlib/fmt/blob/master/include/fmt/ostream.h#L98 doesn't heap-allocate these things, so they wouldn't show up in valgrind. i think we'd need an actual benchmark to measure the difference between a type that uses ostream_formatter and an identical type with a hand-crafted formatter